janus-idp/backstage-plugins
# feat(topology): add permissions to topology plugin

## Description

This PR adds permissions to the Topology plugin. These permissions, `topology.panel.read` and `topology.view.read`, to restrict access to the Topology plugin. Permission `topology.view.read` will limit access to the Topology view, see below. Permission `topology.panel.read` will limit access to the Topology sidebar.

# Fixes

- Fixes: [RHIDP-1794](https://issues.redhat.com/browse/RHIDP-1794)

## Screenshots

`topology.view.read`
![Screenshot from 2024-05-15 11-16-17](https://github.com/janus-idp/backstage-plugins/assets/82788581/5ee18543-51a2-45da-b0fc-ba29a057cf4d)
allow

![Screenshot from 2024-05-15 11-16-30](https://github.com/janus-idp/backstage-plugins/assets/82788581/f7262883-870f-413c-a6d5-99ecca253847)
deny (used `TopologyEmptyState` because the approach is similar to `catalog.entity.read` showing an empty state when navigating to the catalog when set to deny)

`topology.panel.read`
![Screenshot from 2024-05-15 11-17-33](https://github.com/janus-idp/backstage-plugins/assets/82788581/8df3613c-ba65-4f50-bf10-63d3c31ffe57)
allow (details tab)

![Screenshot from 2024-05-15 11-17-35](https://github.com/janus-idp/backstage-plugins/assets/82788581/45b9be45-752c-4e16-a7f4-f6fa8843bf5e)
allow (resources tab)

![Screenshot from 2024-05-15 11-16-49](https://github.com/janus-idp/backstage-plugins/assets/82788581/b9c5e196-f8d2-4ec9-9e1f-3eb687b7cf3d)
deny (tabs are disabled) (went with the `EmptyState` from patternfly and passed in the text `Unauthorized: Unable to view data`, didn't know what text to include and whether or not we should use the `TopologyEmptyState`)

`catalog.entity.read`
![Screenshot from 2024-05-15 11-17-33](https://github.com/janus-idp/backstage-plugins/assets/82788581/c759084f-78ea-409f-bea6-2259aff603c3)
allow

![Screenshot from 2024-05-15 11-17-49](https://github.com/janus-idp/backstage-plugins/assets/82788581/03772f4c-b30b-463d-b6b6-df9f3a8f9fcc)
deny (before reload of browser)

![Screenshot from 2024-05-15 11-20-53](https://github.com/janus-idp/backstage-plugins/assets/82788581/98542cef-9968-4b2a-9366-9a87c4e03791)
deny (after reload of browser, view is originating from catalog plugin)

## Special notes for the Review

### How to test

1. Configure the Topology plugin, Kubernetes plugin, Kubernetes Backend plugin
2. Create a policy csv file for the permission policies
    ```csv
    g, user:default/<YOUR_USERNAME>, role:default/topology

    p, role:default/topology, topology.panel.read, read, allow
    p, role:default/topology, topology.view.read, read, allow
    p, role:default/topology, catalog-entity, read, allow
    ```
3. Configure the RBAC Backend plugin
    ```yaml
    permission:
      enabled: true
      rbac:
        policies-csv-file: ../<PATH>/<TO>/<CSV>/rbac-policy.csv
        policyFileReload: true
    ```
4. Ensure auth is configured
5. Ensure that there is a cluster to connect to, a deployment and catalog entity that can be referenced, and a user that you can sign into in the catalog
6. View the Topology side for you deployed catalog entity
7. Change above permission policies to denied and reload to restrict access

This is marked as a draft to get review and sign off on UI / UX / and the RBAC Backend team. I took the approach of getting everything implemented prior to getting review on how we want it to look because of the priority. Most of the changes are probably where we want them. The biggest things that need review are the text and whether we want to use an empty state or something more customized.

@ShiranHi @invincibleJai @divyanshiGupta
